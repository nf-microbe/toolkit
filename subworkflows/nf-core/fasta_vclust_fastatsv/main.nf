// Import modules
include { SEQHASHER_SEQUNIQ } from '../../../modules/nf-core/seqhasher/sequniq'
include { SEQKIT_GREP       } from '../../../modules/nf-core/seqkit/grep'
include { VCLUST_ALIGN      } from '../../../modules/nf-core/vclust/align'
include { VCLUST_CLUSTER    } from '../../../modules/nf-core/vclust/cluster'
include { VCLUST_PREFILTER  } from '../../../modules/nf-core/vclust/prefilter'

workflow FASTA_VCLUST_FASTATSV {
    take:
    fasta_gz    // [ [ meta ], fasta.gz ]
    vclust      // path/to/vclust/binary

    main:
    ch_versions = Channel.empty()

    //
    // MODULE: Prefilter alignments
    //
    VCLUST_PREFILTER(
        fasta_gz,
        vclust
    )
    ch_versions = ch_versions.mix(VCLUST_PREFILTER.out.versions)

    // combine input by meta
    ch_vclust_align_input = fasta_gz
        .combine(VCLUST_PREFILTER.out.prefilter, by:0)
        .multiMap { meta, fasta, prefilter ->
            fasta:      [ meta, fasta ]
            prefilter:  [ meta, prefilter ]
        }

    //
    // MODULE: Align sequences
    //
    VCLUST_ALIGN(
        ch_vclust_align_input.fasta,
        ch_vclust_align_input.prefilter,
        vclust
    )
    ch_versions = ch_versions.mix(VCLUST_ALIGN.out.versions)

    //
    // MODULE: Cluster sequences
    //
    VCLUST_CLUSTER(
        VCLUST_ALIGN.out.ani,
        VCLUST_ALIGN.out.ids,
        vclust
    )
    ch_versions = ch_versions.mix(VCLUST_CLUSTER.out.versions)

    //
    // MODULE: Dereplicate cluster output to retain only cluster rep
    //
    SEQHASHER_SEQUNIQ(
        VCLUST_CLUSTER.out.clusters
    )
    ch_versions = ch_versions.mix(SEQHASHER_SEQUNIQ.out.versions)

    // combine input by meta
    ch_seqkit_grep_input = fasta_gz
        .combine(SEQHASHER_SEQUNIQ.out.unique, by:0)
        .multiMap { meta, fasta, sequniq ->
            fasta:      [ meta, fasta ]
            sequniq:    [ meta, sequniq ]
        }

    //
    // MODULE: Extract rep sequences from FastA file
    //
    SEQKIT_GREP(
        ch_seqkit_grep_input.fasta,
        ch_seqkit_grep_input.sequniq.map { meta, tsv -> tsv }
    )
    ch_versions = ch_versions.mix(SEQKIT_GREP.out.versions)

    emit:
    cluster_reps_fasta_gz   = SEQKIT_GREP.out.filter    // [ [ meta ], fasta.gz ]
    versions                = ch_versions               // channel: [ versions.yml ]
}
