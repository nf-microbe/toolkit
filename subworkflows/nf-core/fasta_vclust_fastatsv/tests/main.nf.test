nextflow_workflow {

    name "Test Subworkflow FASTA_VCLUST_FASTATSV"
    script "../main.nf"
    workflow "FASTA_VCLUST_FASTATSV"
    config "./nextflow.config"

    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/fasta_vclust_fastatsv"
    tag "seqhasher"
    tag "seqhasher/sequniq"
    tag "seqkit"
    tag "seqkit/grep"
    tag "vclust"
    tag "vclust/align"
    tag "vclust/cluster"
    tag "vclust/prefilter"

    test("b_fragilis - fasta.gz") {
        when {
            params {
                vclust_cluster_args = "--algorithm complete --metric ani --ani 0.95 --qcov 0.5 --out-repr"
            }
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ id:'test1', single_end:false ], // meta map
                        file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/illumina/fasta/test1.contigs.fa.gz', checkIfExists: true)
                    ],
                    [
                        [ id:'test2', single_end:false ], // meta map
                        file("${projectDir}/assets/test-datasets/genomics/prokaryotes/bacteroides_fragilis/illumina/fasta/test2.contigs.fa.gz", checkIfExists: true)
                    ]
                )
                input[1] = file("${projectDir}/bin/vclust", checkIfExists: true)
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.versions
                    ).match()
                }
            )
        }
    }
}
