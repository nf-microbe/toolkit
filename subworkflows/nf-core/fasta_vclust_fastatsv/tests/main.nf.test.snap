{
    "b_fragilis - fasta.gz": {
        "content": [
            [
                "versions.yml:md5,254cee6b6b2a35094c27e745590e1d43",
                "versions.yml:md5,254cee6b6b2a35094c27e745590e1d43",
                "versions.yml:md5,a5ea7279e18f784bbdc6a1a947f53735",
                "versions.yml:md5,a5ea7279e18f784bbdc6a1a947f53735",
                "versions.yml:md5,b32b2b59f4acde593188a1ade7f1b371",
                "versions.yml:md5,b32b2b59f4acde593188a1ade7f1b371",
                "versions.yml:md5,c5017d798be666740adc0950007bfb8c",
                "versions.yml:md5,c5017d798be666740adc0950007bfb8c",
                "versions.yml:md5,dbca785f9f6f2d19599f052ab9612ade",
                "versions.yml:md5,dbca785f9f6f2d19599f052ab9612ade"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-15T13:49:49.927871971"
    }
}